# 2024_GonzalezOjeda_CLARC

## About

Code corresponding to the paper: 

Linkage-based ortholog refinement in bacterial pangenomes with CLARC (2024)

Indra González Ojeda, Samantha G. Palace, Pamela P. Martinez, Taj Azarian, Lindsay R. Grant, Laura L. Hammitt, Katherine L. O’Brien, William P. Hanage, Marc Lipsitch

## Table of contents

- [Genomes](#genomes)
- [Figure 1](#figure-1)
- [Figure 2](#figure-2)
- [Figure 3](#figure-3)
- [Figure 4](#figure-4)
- [Figure 5](#figure-5)
- [Figure 6](#figure-6)
- [Figure 7](#figure-7)
- [Guide to replicating the figures](#guide-to-replicating-the-figures)

## Genomes

A total of 8907 pneumococcal genomes were used in this study. Among these are 9 closed references and samples from 7 different carriage datasets. The locations and number of samples used from each dataset are:

- Southwest, USA - 937 samples
- Massachusetts, USA - 1347 samples
- Southampton, UK - 470 samples
- Maela, Thailand - 2920 samples
- Malawi - 629 samples
- South Africa - 1637 samples
- Iceland - 958 samples

The accession numbers for all samples used (divided by dataset) can be found in the [Genomes/accession_numbers](https://github.com/IndraGonz/2024_GonzalezOjeda_CLARC/tree/main/Genomes/accession_numbers/) folder within this repository. 

Additionally, all corresponding assemblies (.fasta) and annotation files (.gff) can be found in this [Zenodo folder]()
  
## Figure 1

Workflow of CLARC tool:

<img src="https://github.com/IndraGonz/2024_GonzalezOjeda_CLARC/blob/main/Figures/images/Figure1_v3.png" alt="CLARC workflow" width="750"/>  

## Figure 2

Distribution of pneumococcal strain clusters in different geographic locations:

<img src="https://github.com/IndraGonz/2024_GonzalezOjeda_CLARC/blob/main/Figures/images/Figure2_v3.png" alt="Figure 2" width="900"/>  

### 2B, 2C

- code:
  - Jupyter notebook with processing of PopPUNK data: [Figure2_code_CLARC_2024.ipynb](https://github.com/IndraGonz/2024_GonzalezOjeda_CLARC/blob/main/Figures/Fig2/code/Figure2_code_CLARC_2024.ipynb)
 
- data:
  - Output of PopPUNK analyses, along with the appropiate qfiles for each run can be found at: [Figures/Fig2/data/poppunk_typing](https://github.com/IndraGonz/2024_GonzalezOjeda_CLARC/tree/main/Figures/Fig2/data/poppunk_typing)

## Figure 3

## Figure 4

## Figure 5

## Figure 6

## Figure 7

## Guide to replicating the figures
